rotkehlchen/assets/utils.py

```diff
@@ -68,36 +68,43 @@ def get_or_create_evm_token(
-    except (UnknownAsset, DeserializationError):
-        log.info(
-            f'Encountered unknown asset with address '
-            f'{evm_address}. Adding it to the global DB',
-        )
-        if ethereum_manager is not None:
-            info = ethereum_manager.get_basic_contract_info(evm_address)
-            decimals = info['decimals'] if decimals is None else decimals
-            symbol = info['symbol'] if symbol is None else symbol
-            name = info['name'] if name is None else name
-
-            if None in (decimals, symbol, name):
-                raise NotERC20Conformant(f'Token {evm_address} is not ERC20 conformant')  # noqa: E501  # pylint: disable=raise-missing-from
-
-        token_data = EvmToken.initialize(
-            address=evm_address,
-            chain=chain,
-            token_kind=token_kind,
-            name=name,
-            decimals=decimals,
-            symbol=symbol,
-            protocol=protocol,
-            underlying_tokens=underlying_tokens,
-        )
-        # This can but should not raise InputError since it should not already exist
-        ethereum_token = add_ethereum_token_to_db(token_data)
-        with userdb.user_write() as cursor:
-            userdb.add_asset_identifiers(cursor, [ethereum_token.identifier])
+    # Protect the whole section with a lock to avoid the case where the try is executed
+    # by two greenlets, both enter the except and then they fail. The existence check and addition
+    # is a critic section.
+    with userdb.token_addition_lock:
+        try:
+            ethereum_token = EvmToken(
+                identifier=identifier,
+                form_with_incomplete_data=form_with_incomplete_data,
+            )
+        except (UnknownAsset, DeserializationError):
+            log.info(
+                f'Encountered unknown asset with address '
+                f'{evm_address}. Adding it to the global DB',
+            )
+            if ethereum_manager is not None:
+                info = ethereum_manager.get_basic_contract_info(evm_address)
+                decimals = info['decimals'] if decimals is None else decimals
+                symbol = info['symbol'] if symbol is None else symbol
+                name = info['name'] if name is None else name
+
+                if None in (decimals, symbol, name):
+                    raise NotERC20Conformant(f'Token {evm_address} is not ERC20 conformant')  # noqa: E501  # pylint: disable=raise-missing-from
+
+            token_data = EvmToken.initialize(
+                address=evm_address,
+                chain=chain,
+                token_kind=token_kind,
+                name=name,
+                decimals=decimals,
+                symbol=symbol,
+                protocol=protocol,
+                underlying_tokens=underlying_tokens,
+            )
+            # This can but should not raise InputError since it should not already exist
+            ethereum_token = add_ethereum_token_to_db(token_data)
```

Comment:
My comment here is that as far as I can see the only place in this function when context switch can happen is these 2 DB functions. (I don't think that logging functions can context switch, right?)

So instead of adding a new lock you can perhaps re-introduce (seems it's removed?) the critical section in the DB cursor creation that puts `set_progress_handler` to `0` and resets it to the last value once out of the section?

This approach may remove the need for the extra lock.

So I mean to reverts this commit: 339244579bec8c2c97d37ac01a2c8190f89bd3a7 except for the actual usage. So except for using `self.enter_critical_section()` and `self.exit_critical_section()`.

Instead you can use those here in this function by making a different ctx() method for critical section and using that ctx() in this function that enters and exits critical section.

Does that make sense?

You would need to do that for both the global db write at `add_ethereum_token_to_db` and userdb write with `add_asset_identifiers`.

Reply:
I've doing a bit of research and seems that it possible to make the logs async using https://github.com/gevent/gevent/blob/beb98ce4742181caaa6cec9cedecaaa2a04f910b/src/gevent/os.py#L28 but is not the default afaik and we don't do that.

Reply:
but I believe your proposal doesn't solve the issue because the problem is not writing at the same time in the database. The problem is that the two operations of checking that a token exists or not and the process of adding it to the database if not shouldn't be executed by two greenlets at the same time. Both the read and the write should be "atomic"

Reply:
I am not saying to make the logs async.

I am saying that so long as there is nothing that will context switch the entire function out of the current greenlet then the function will indeed in fact be atomic.

Reply:
Okey got you. My mind was thinking in threads and not in greenlets. I believe is correct now